### PR TITLE
async client: refactor `Http::AsyncClient::Callbacks` to pass in `Http::AsyncClient::Request*` for correlation between requests and responses

### PR DESCRIPTION
**Pull request by yskopets**
Description: refactor `Http::AsyncClient::Callbacks` to pass in `Http::AsyncClient::Request*` for correlation between requests and responses
Risk Level: Low
Testing: unit tests
Docs Changes: N/A
Release Notes: N/A

Context:
* related to #9998 

Motivation:
* let `Callbacks` implementations to be able to determine which of the multiple in-flight requests has changed it's status

### CODE REVIEW

**Comment by alyssawilk** [comment]
cc @alyssawilk @mattklein123

File: include/envoy/http/async_client.h [hunk 1]
```diff
@@ -30,11 +43,42 @@ class AsyncClient {
...
+     *                requests in progress.
+     * @param reason  failure reason
+     */
+    virtual void onRequestFailure(const Request* request, FailureReason reason) PURE;
+  };
+
+  /**
+   * Notifies caller of async HTTP request status.
+   */
+  class RequestCallbacks : public Callbacks {
```
Comment (snowp) [inline include/envoy/http/async_client.h]: I'm not sure if having two variations of this interface is worth it, would it be possible to just update the existing implementations to override the function with the request ptr parameter?

File: source/common/http/async_client_impl.cc [hunk 2]
```diff
@@ -267,7 +267,7 @@ void AsyncRequestImpl::onComplete() {
                                                    response_->trailers(), streamInfo(),
                                                    Tracing::EgressConfig::get());
 
-  callbacks_.onSuccess(std::move(response_));
+  callbacks_.onRequestSuccess(this, std::move(response_));
```
Comment (snowp) [inline source/common/http/async_client_impl.cc]: if we're always passing `this` i don't think it can be null? should we be passing a reference instead?

Comment (yskopets) [reply source/common/http/async_client_impl.cc]: fixed

File: include/envoy/http/async_client.h [hunk 3]
```diff
@@ -43,8 +87,17 @@ class AsyncClient {
...
     /**
      * Called when the async HTTP request fails.
+     * @param reason failure reason
      */
     virtual void onFailure(FailureReason reason) PURE;
+
+    // Callbacks
+
+    void onRequestSuccess(const Request*, ResponseMessagePtr&& response) override {
+      onSuccess(std::forward<ResponseMessagePtr>(response));
```
[the remark this answers is not in the document]
Comment (yskopets) [reply include/envoy/http/async_client.h]: replacing `std::forward` with `std:move` results in a compilation error:

```
bazel-out/darwin-fastbuild/bin/include/envoy/http/_virtual_includes/async_client_interface/envoy/http/async_client.h:97:17: error: no matching function for call to 'move'
      onSuccess(std::move<ResponseMessagePtr>(response));
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/type_traits:2310:1: note: candidate function template not viable: cannot convert argument of incomplete type 'Envoy::Http::ResponseMessagePtr' (aka 'unique_ptr<Message<Envoy::Http::ResponseHeaderMap, Envoy::Http::ResponseTrailerMap> >') to 'std::__1::unique_ptr<Envoy::Http::Message<Envoy::Http::ResponseHeaderMap, Envoy::Http::ResponseTrailerMap>, std::__1::default_delete<Envoy::Http::Message<Envoy::Http::ResponseHeaderMap, Envoy::Http::ResponseTrailerMap> > > &&' for 1st argument
move(_Tp&& __t) _NOEXCEPT
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/iterator:1498:48: note: candidate function template not viable: requires 3 arguments, but 1 was provided
    template <class _Ip, class _Op> friend _Op move(_Ip, _Ip, _Op);
                                               ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__bit_reference:692:1: note: candidate function template not viable: requires 3 arguments, but 1 was provided
move(__bit_iterator<_Cp, _IsConst> __first, __bit_iterator<_Cp, _IsConst> __last, __bit_iterator<_Cp, false> __result)
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/deque:735:1: note: candidate function template not viable: requires at least 3 arguments, but 1 was provided
move(_RAIter __f,
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/deque:765:1: note: candidate function template not viable: requires 3 arguments, but 1 was provided
move(__deque_iterator<_V1, _P1, _R1, _M1, _D1, _B1> __f,
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/deque:793:1: note: candidate function template not viable: requires 3 arguments, but 1 was provided
move(__deque_iterator<_V1, _P1, _R1, _M1, _D1, _B1> __f,
^
1 error generated.
```

File: include/envoy/http/async_client.h (hunk 1 above)
Comment (yskopets) [reply include/envoy/http/async_client.h]: `RequestCallbacks` has been removed

**Comment by yskopets** [comment]
@alyssawilk No, sorry, I don't have such a test.

I'll see what I can do.

**Comment by yskopets** [comment]
@alyssawilk I've added missing unit tests.

File: test/common/config/datasource_test.cc [hunk 4]
```diff
@@ -115,7 +116,7 @@ TEST_F(AsyncDataSourceTest, LoadRemoteDataSourceReturnFailure) {
 
   initialize([&](Http::RequestMessagePtr&, Http::AsyncClient::Callbacks& callbacks,
                  const Http::AsyncClient::RequestOptions&) -> Http::AsyncClient::Request* {
-    callbacks.onFailure(Envoy::Http::AsyncClient::FailureReason::Reset);
+    callbacks.onFailure(request_, Envoy::Http::AsyncClient::FailureReason::Reset);
     return nullptr;
```
Comment (snowp) [inline test/common/config/datasource_test.cc]: should you be returning the request_ ptr here?

Comment (yskopets) [reply test/common/config/datasource_test.cc]: Actually, I've decided to restore original `return nullptr`  in all other places.

`return nullptr` affects the code path that follows `AsyncClient::send()`, so I shouldn't have changed it in the first place.

**Pull request merged**